### PR TITLE
Bring loki tests in line, too

### PR DESCRIPTION
**Pull request by rbarry82**
While we're at it, unpin the operator hash.

Fix the unit test mocking to reflect the actual charm flow.

Emit the configure() hook when checking whether or not it can
connect or not.

Most importantly, FIX _provide_loki so it doesn't try to bring
up a second Provider if one is already set, otherwise
`ops.framework` sees two objects with the same handle and throws
an exception back at us.

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -116,54 +120,51 @@ def test__alerting_config(self):
...
-        self.harness.charm._container.can_connect.return_value = False
-        self.harness.update_config(yaml.safe_load(LOKI_CONFIG))
+
+        # Since harness was not started with begin_with_initial_hooks(), this must
+        # be emitted by hand to actually trigger _configure()
+        self.harness.charm.on.config_changed.emit()
         self.assertIsInstance(self.harness.charm.unit.status, WaitingStatus)
 
     @patch("ops.testing._TestingPebbleClient.push")
+    @patch("ops.testing._TestingPebbleClient.make_dir")
```
Comment (sed-i) [inline tests/unit/test_charm.py:135]: Mind removing these?

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -116,54 +120,51 @@ def test__alerting_config(self):
...
-        self.harness.charm._container.can_connect.return_value = True
-        self.harness.update_config(yaml.safe_load(LOKI_CONFIG))
+        self.harness.set_leader(True)
+
+        # Since harness was not started with begin_with_initial_hooks(), this must
+        # be emitted by hand to actually trigger _configure()
+        self.harness.charm.on.config_changed.emit()
         self.assertIsInstance(self.harness.charm.unit.status, ActiveStatus)
 
     @patch("ops.testing._TestingPebbleClient.make_dir")
```
Comment (sed-i) [inline tests/unit/test_charm.py:147]: Remove?

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -116,54 +120,51 @@ def test__alerting_config(self):
...
 
-    @patch("loki_server.LokiServer.version", new_callable=PropertyMock)
-    def test__provide_loki_server_error(self, mock_version):
-        mock_version.side_effect = LokiServerError
+    def test__provide_loki_server_error(self):
+        self.mock_version.side_effect = LokiServerError
         self.harness.charm._provide_loki()
         self.assertIsInstance(self.harness.charm.unit.status, BlockedStatus)
 
     @patch("ops.testing._TestingPebbleClient.make_dir")
```
Comment (sed-i) [inline tests/unit/test_charm.py:166]: Remove?

File: tox.ini [hunk 4]
```diff
@@ -39,82 +40,64 @@ commands =
...
 
 [testenv:integration]
 description = Run integration tests
 deps =
+    aiohttp
     git+https://github.com/juju/python-libjuju.git
     #juju
     pytest
     git+https://github.com/charmed-kubernetes/pytest-operator.git
+    tenacity
```
Comment (sed-i) [inline tox.ini:101]: aiohttp and tenacity needed?

**Review by sed-i — APPROVED** [review]

File: tests/unit/test_charm.py (hunk 1 above)
Comment (rbarry82) [reply tests/unit/test_charm.py:135]: `make_dir`, sure. `push` is actually needed.

File: tox.ini (hunk 4 above)
Comment (rbarry82) [reply tox.ini:101]: Tenacity is not. aiohttp will be whenever integration tests come in if we ever want to use the [Loki API](https://grafana.com/docs/loki/latest/api/) in integration tests, so may as well keep it IME

**Review by dstathis — APPROVED** [review]

**Review by sed-i — DISMISSED** [review]

File: .github/workflows/codeql-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    # ✏️ If the Autobuild fails above, remove it and uncomment the following three lines
+    #    and modify them (or add more) to build your code if your project
+    #    uses a compiled language
+
+    #- run: |
+    #   make bootstrap
+    #   make release
+
+    - name: Perform CodeQL Analysis
+      uses: github/codeql-action/analyze@v1
```
Comment (simskij) [inline .github/workflows/codeql-analysis.yml:64]:
```suggestion
name: "CodeQL"

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

permissions:
  security-events:
    write

jobs:
  analyze:
    name: Analyze
    runs-on: ubuntu-latest
    strategy:
      fail-fast: false
      matrix:
        language: ['python']
    steps:
    - name: Checkout repository
      uses: actions/checkout@v2

    - name: Initialize CodeQL
      uses: github/codeql-action/init@v1
      with:
        languages: ${{ matrix.language }}
        
    - name: Autobuild
      uses: github/codeql-action/autobuild@v1
      
    - name: Perform CodeQL Analysis
      uses: github/codeql-action/analyze@v1
```

**Review by simskij — APPROVED** [review]
Some minor comments. Up to you if you want to address them or not, the PR is approved regardless. 👍🏽

File: tox.ini [hunk 6]
```diff
@@ -39,82 +40,63 @@ commands =
...
     coverage[toml]
+    responses
+    -r{toxinidir}/requirements.txt
     deepdiff
 commands =
     coverage run \
       --source={[vars]src_path},{[vars]lib_path} \
-      -m pytest -v -s --tb native {posargs} {[vars]test_path}/unit
-    coverage report -m
+      -m pytest -v --tb native -s {posargs} {[vars]tst_path}/unit
```
[the remark this answers is not in the document]
Comment (rbarry82) [reply tox.ini:90]: Intentional (everywhere in the file), since `tst_path` is what's used by other repos, it's become the de-facto standard

**Comment by rbarry82** [comment]
Of course it needs a re-review after merging @simskij's suggestion

**Comment by simskij** [comment]
> Of course it needs a re-review after merging @simskij's suggestion

Sometimes, when the change is insignificant enough, I'm able to do some magic 😜

**Pull request merged**